Fix C++20/gcc-12 issues (Part 2)

falbrechtskirchinger opened the pull request:
This is part 2 of #3379.

## Add the spaceship operator

Part 1 had to disable some comparisons in unit tests because of incorrect results due to standard library code using 3-way comparison which, in the absence of a user-defined operator, compares the result of implicit conversions.
In part 2 we implement the spaceship operator and fix some inconsistencies in the old comparison code:
* `NaN` values always compare `false`/unordered when compared to themselves, other `NaN`s, or other numbers.
* Discarded values will always compare `false`/unordered to themselves or any other value, unless the legacy comparison behavior is explicitly enabled by defining the macro `JSON_LEGACY_COMPARISON=1`.
When the legacy behavior is selected, operations that were computed as an odd number of inverses of other operations always return `true`. Example:
    Given two values `lhs` and `rhs` and at least one of them being a discarded value.
    1. (lhs < rhs) == false
    2. (lhs >= rhs) == true, because it is computed as !(lhs < rhs) (odd)
    3. (lhs > rhs) == false, because it is computed as !(lhs <= rhs) which is computed as !(rhs < lhs) (even)

GCC ignores user-defined `operator<=>` for enumerated types when rewriting operators. Clang, MSVC, and ICC do not. (ICC manages to produce an incorrect result anyway o.O.)
I've filed a bug with GCC but do agree now that GCC is following the letter of the standard. If anyone knows where to best raise concerns over a language defect, let me know.
it's easy enough to work around by defining `operator<` and explicitly delegating to `operator<=>`. Currently only enabled for GCC but almost guaranteed to be required for (some versions of) ICC.

The unit tests were updated in several ways:
* Defined equal-length constants for `bool` and `std::partial::ordering`.
  The goal was to find visually distinguishable names for `true` and `false`. `T`/`F`, `t`/`f`, `tr`/`fa` were all ruled out because they look too similar (in my editor anyway). `f_` and `_t` aren't perfect but the best I've come up with so far.
* Replaced constants, aligned all tables, and numbered rows and columns.
* Added `value_t::discarded` to the list of types and updated tables.
* Added `NaN` and `json(value_t::discarded)` to the list of values and updated tables.
* Added tables for the results of 3-way comparisons.
* Cross-check comparisons with their 3-way comparison result and perform range check between both tables on traversal.
* Added `test-comparison_legacy` based on `unit-comparison.cpp` but with `JSON_LEGACY_COMPARISON=1`.
* Added a regression test to verify a unit test from `unit-class_parser.cpp`, which breaks with the new comparison behavior, continues to work in legacy mode.

## Fix iterators to meet (more) `std::ranges` requirements

Unlike the unconstrained algorithms, which only declared named requirements, constrained algorithms and views use concepts to enforce requirements.
Some iterators did not meet enough of those requirements.
In this PR we ensure that `iter_impl` satisfies the `std::bidirectional_iterator` concept and `iteration_proxy_value` satisfies the `std::input_iterator` concept.

The `apply*()` functions included in an earlier revision were moved into a separate PR: #3450

<hr />

To Do:

- [x] Check iterator concepts in unit test.
- [x] Add note about "magic keywords" to unit tests.
- [X] Should `JSON_LEGACY_COMPARISON=1` be added to CI?
- [X] Should `JSON_LEGACY_COMPARISON` be renamed to `JSON_LEGACY_DISCARDED_COMPARISON` (or something else entirely) now that it only applies to discarded values?
- [x] Document macro `JSON_LEGACY_COMPARISON`.
- [x] Document macro `JSON_HAS_THREE_WAY_COMPARISON`.
- [x] Document macro `JSON_HAS_RANGES`.
- [x] Update operator documentation.
- [X] Complete issue references in commit messages.

<hr />

Fixes #3130.
Fixes #3409.

Related discussion: #3408

gregmarr commented on include/nlohmann/detail/iterators/iteration_proxy.hpp:
```diff
@@ -40,15 +44,23 @@ template<typename IteratorType> class iteration_proxy_value
     mutable string_type array_index_str = "0";
     /// an empty string (to return a reference for primitive values)
-    const string_type empty_str{};
+    string_type empty_str{};
```
Why is this no longer const?

falbrechtskirchinger replied:
To get around the issue of not being able to default the copy assignment operator. I thought that was the simplest solution of all. `static` isn't one unless we disable the exit-time destructor warnings.

gregmarr replied:
Ah, right.  That's annoying.  It means that we don't get the compiler protection against something accidentally modifying it.

> static isn't one unless we disable the exit-time destructor warnings.

What is this referring to?

falbrechtskirchinger replied:
> > static isn't one unless we disable the exit-time destructor warnings.
> 
> What is this referring to?

It's referring to me having had a bad time using statics in the abandoned "Improve unit testing (2)" PR. Long story short, statics of non-trivial type involve destructor calls on program exit, that can cause issues so it's good practice to not use them and also turn on warning flags for them (and they are for this project). There are workarounds but i'd rather stay clear of them.

gregmarr replied:
Okay.  I wonder if just writing out the copy assignment operator by hand is still the best solution.

gregmarr replied:
Yes, it can make sure that we don't accidentally write code that modifies it at some point in the future.  I think that there are probably optimizations too, but that's less of a concern.

falbrechtskirchinger commented:
Sadly, `iter_impl` cannot be a random access iterator. I was hoping that we could just throw exceptions in the random access iterator functions when our underlying iterator isn't a random access iterator.
But functions like `std::distance` select a different implementation at compile-time based on the iterator category.

Right now`std::sort(j.begin(), j.end())` should not work because sorting requires random access. To workaround these limitations, I'm exploring two additional functions (available from C++14 onward; C++11 may be possible with reduced functionality):
```cpp
template<typename Fn, typename... Args>
decltype(auto) apply(Fn&& f, Args&& ... args);

template<typename Fn, typename... Args>
decltype(auto) apply_iter(Fn&& f, Args&& ... args);
```
Those functions would then enable the following:
```cpp
j.apply(std::ranges::sort);
j.apply_iter(std::sort);
```
And I don't know how useful it is but the following would also just happen to work:
```cpp
struct foo {
    int i = 0;
    void set_i(int i_) {
        i = i_;
    }
};

json j(42);
foo f;
j.apply(&f.i); // sets f.i to 42
j.apply(&foo::set_i, f); // same; technically this requires some extra work because of how arguments are ordered
```
(That's all just because of how the standard defines the `INVOKE()` operation.)

<hr/>

Separate issue. Because `iter_impl` now stores a pointer, it's possible to differentiate JSON `null` values. It makes sense to me to forward `nullptr` in `apply()`. For symmetry, `apply_iter()` should do the the same.
```cpp
    reference operator*() const
    {
        JSON_ASSERT(m_object != nullptr);

        switch (m_object->m_type)
        {
            // ...
            case value_t::null:
                JSON_THROW(invalid_iterator::create(214, "cannot get value", m_object));
```
Can we get rid of that exception and return `*m_object` instead? How bad of a breaking change would that be?

falbrechtskirchinger commented:
Or … just change the failing unit tests to not use `std::distance()` to get the size of an object.

Option A) Add about 200 lines of code for a clumsy, unintuitive way of calling `std::ranges::sort()` on a `basic_json`.
Option B) Accept that `std::distance()` can't be called on objects but `std::ranges::sort(j)` just works.

Option B would be a breaking change for some users. What should we do? @nlohmann

falbrechtskirchinger commented:
Ignore the "Disable regression test for 3070 on GCC <8.4" commit for now. I rebased incorrectly against a local branch instead of upstream/develop. I'll fix it when I push the next set of changes.

falbrechtskirchinger commented on doc/mkdocs/Makefile:
```diff
@@ -2,6 +2,9 @@
 
+serve_dirty: prepare_files
+	venv/bin/mkdocs serve --dirtyreload
+
```
Didn't mean to commit this … but it doesn't hurt, right?

nlohmann replied:
I am just working on the `overwork_documentation` branch and was annoyed by the slow updates. Was not aware of this - yes, please keep it!

falbrechtskirchinger commented on cmake/test.cmake:
```diff
@@ -165,22 +166,30 @@ endfunction()
 #
 # if C++ standard <version_number> is supported by the compiler and the
 # source file contains JSON_HAS_CPP_<version_number>.
+# Use NAME <name> to override the filename-derived test name.
```
This new argument to `json_test_add_test_for()` is used to add the same unit test file multiple times with different build settings.

falbrechtskirchinger commented on doc/mkdocs/docs/api/basic_json/operator_eq.md:
```diff
@@ -1,21 +1,31 @@
-  are the same according to their respective `operator==`.
-- Integer and floating-point numbers are automatically converted before comparison. Note that two NaN values are always
-  treated as unequal.
+2. Compares a JSON value and a scalar or a scalar and a JSON value by converting the scalar to a JSON value and comparing both JSON values (see 1).
```
Note to self: update wording based on operator_ne(?).

nlohmann commented on doc/mkdocs/mkdocs.yml:
```diff
@@ -31,6 +31,7 @@ theme:
 nav:
   - Home:
     - index.md
+    - "API Documentation": api/basic_json/index.md
```
I get why you're doing this, but I think it's confusing to have the menu item twice - especially since the behavior of the other pages is different.

falbrechtskirchinger replied:
I did not plan to keep it there. I noticed the other day, that I'm zoomed in to 150% on json.nlohmann.me which removes the menu bar. I was missing a direct link to the API documentation all this time. :-)

nlohmann commented on doc/mkdocs/docs/css/custom.css:
```diff
@@ -0,0 +1,4 @@
+/* disable ligatures in code and preformatted blocks */
+code, pre {
+    font-variant-ligatures: none;
```
Is this a concern for you? I am using Jetbrains Mono and enjoy their ligatures.

falbrechtskirchinger replied:
Most annoyingly it turns `<=>` into iff (if and only if). The other operators are also subject to ligatures which is slightly less distracting.

nlohmann replied:
I have never perceived it as iff - but I guess now I cannot unsee it :)

nlohmann commented on test/src/unit-class_parser.cpp:
```diff
@@ -33,6 +33,7 @@ SOFTWARE.
 #include <nlohmann/json.hpp>
 using nlohmann::json;
 
+#include <iostream>
```
This looks like a leftover.

falbrechtskirchinger replied:
Indeed.

nlohmann commented on test/src/unit-iterators2.cpp:
```diff
@@ -881,4 +893,100 @@ TEST_CASE("iterators 2")
+    // JSON_HAS_CPP_20 (do not remove; see note at top of file)
+    SECTION("ranges")
+    {
+#if !defined(__clang__) || (defined(__clang__) && defined(__GLIBCXX__))
```
Please add a comment for this compiler filter.

falbrechtskirchinger commented on doc/mkdocs/docs/api/basic_json/operator_spaceship.md:
````diff
@@ -0,0 +1,70 @@
+
+```cpp
+// since C++20
+class basic_json {
````
I'm obviously not putting down the complete template definition here but in other places I've abbreviated the template with `template<...>`.

Should I simply leave it off everywhere? It should be clear enough from context that `basic_json` refers to the main template class.

nlohmann replied:
In the other API pages, we do not repeat the class `basic_json` - it's only hinted in the title. I just copied this way from cppreference.com - or do I misunderstand?

falbrechtskirchinger replied:
Well, previously operators where non-member friend functions. In C++20 they become member functions. This particular page (about `operator<=>`) doesn't exist prior to C++20 so it's not needed here. But the problem still exists on the other operator pages.

nlohmann replied:
cppreference is doing it like this:

- member function: https://en.cppreference.com/w/cpp/string/basic_string/operator%2B%3D

- nonmember function: https://en.cppreference.com/w/cpp/string/basic_string/operator%2B

falbrechtskirchinger replied:
But we have them on the same page. Going from non-member to member. That's why I added the `class basic_json { }` explicitly.

nlohmann replied:
I see. Let's keep it this way until we find a better way. I like how cppreference is a bit more colorful...

falbrechtskirchinger commented on doc/mkdocs/docs/api/basic_json/operator_lt.md:
```diff
@@ -1,31 +1,34 @@
+        5. array
+        6. string
+        7. binary
+      For instance, any boolean value is considered less than any string.
```
Technically, this is now duplicated in `value_t.md` but linking there and leaving the example here makes no sense and moving the example (as is) to `value_t.md` removes needed context.

nlohmann commented on doc/mkdocs/docs/api/macros/json_has_ranges.md:
````diff
@@ -0,0 +1,23 @@
+
+Unless defined by the user prior to inclusion of the `json.hpp` header, the macro definition is roughly equivalent to the follwing:
+```teratermmacro
+#if defined(__cpp_lib_ranges)
````
I would not give a complete definition here. For `JSON_HAS_FILESYSTEM`/`JSON_HAS_EXPERIMENTAL_FILESYSTEM`, I wrote:

> The default value is detected based on the preprocessor macros `#!cpp __cpp_lib_filesystem`,
`#!cpp __cpp_lib_experimental_filesystem`, `#!cpp __has_include(<filesystem>)`, or
`#!cpp __has_include(<experimental/filesystem>)`.

nlohmann commented on doc/mkdocs/docs/api/macros/json_has_ranges.md:
````diff
@@ -0,0 +1,23 @@
+## Default definition
+
+Unless defined by the user prior to inclusion of the `json.hpp` header, the macro definition is roughly equivalent to the follwing:
+```teratermmacro
````
What is `teratermmacro`?

falbrechtskirchinger replied:
That's a mistake I already noted. I was experimenting with syntax highlighting.

falbrechtskirchinger commented on doc/mkdocs/docs/api/macros/json_has_ranges.md:
````diff
@@ -0,0 +1,23 @@
+## Default definition
+
+Unless defined by the user prior to inclusion of the `json.hpp` header, the macro definition is roughly equivalent to the follwing:
+```teratermmacro
````
Whoops. I was experimenting with syntax highlighting. Those should all be `cpp`. Sadly the highlighting with `pygments` is quite monotone.

nlohmann commented on doc/mkdocs/docs/api/macros/json_has_three_way_comparison.md:
```diff
@@ -0,0 +1,24 @@
+
+## Default definition
+
+Unless defined by the user prior to inclusion of the `json.hpp` header, the macro definition is roughly equivalent to the follwing:
```
See above - do not give a complete definition here - the interested reader will like the code more anyway. :)

falbrechtskirchinger commented on doc/mkdocs/docs/api/macros/json_legacy_comparison.md:
```diff
@@ -0,0 +1,51 @@
+| `>=`         | `#!cpp true`  |
+| `>`          | `#!cpp false` |
+
+Otherwise, comparisons involving involving at least one discarded JSON value always yield `#!cpp false`.
```
Typo

falbrechtskirchinger commented on doc/mkdocs/docs/api/macros/json_has_ranges.md:
````diff
@@ -0,0 +1,23 @@
+#define JSON_HAS_RANGES /* value */
+```
+
+Wether the standard library has any support for ranges. Implies support for concepts.
````
Typo and "This macro indicates whether ...".

nlohmann commented on doc/mkdocs/docs/api/macros/json_legacy_comparison.md:
```diff
@@ -0,0 +1,51 @@
+    Be advised that standard library and third-party code can still use the 3-way comparison
+    operator directly, resulting in inconsistent and unpredictable behavior.
+
+!!! warning "Deprecation"
```
I am not 100% happy with the term "deprecation" here. In fact, 3.11.0 will be a breaking change, and this macro is the only way to get back the legacy behavior. I don't find better words though...

falbrechtskirchinger replied:
The idea would be that the pre-3.11.0 behavior, i.e. legacy behavior, can be restored until it is removed in 4.0.0. Hence the macro is deprecated?

falbrechtskirchinger replied:
I have rearranged and reworded the section.

nlohmann replied:
It's better, thanks!

The more I think of it, maybe (not in this PR, but for 3.11.0) we should add a dedicated page for comparison of JSON values where, on one page, the core ideas are described. This would also be a place to address https://github.com/nlohmann/json/issues/3443.

falbrechtskirchinger replied:
A folder and category to place general help topics in would also be useful. I've repeatedly discarded help topics because I found no place to put them.

For example, @gregmarr suggested a page on ADL.

nlohmann replied:
Right. I like the idea of https://documentation.divio.com but did not find the time to reorganize the documentation. Right now, I think just creating one page per topic is fine. I think it helps everybody:

- I often found bugs when I was forced to explain a concept in detail. Or I forgot an important requirement while writing a small example.
- I can reuse the docs very soon in issues or discussions, so it's an investment at least for me.
- A lot of questions never occur, because people can help themselves.
- mkdocs is providing a good-enough search, so even before cleaning up nothing is really "hidden".

nlohmann commented on doc/mkdocs/docs/api/macros/json_legacy_comparison.md:
```diff
@@ -0,0 +1,51 @@
+
+!!! warning "Inconsistent behavior in C++20-mode"
+
+    The `<=>` operator will always yield the correct result (`#!cpp false`) but overloads
```
Does `<=>` ignore `JSON_LEGACY_COMPARISON`?

falbrechtskirchinger replied:
Yes. I did not want to hack the legacy behavior into the new, clean comparison function. It would also make an already slight more expensive operation even more expensive, since I'd have to do some upfront checks first, then do the already more expensive 3-way comparison.

nlohmann reviewed (commented):
Since I keep making comments to things you are then commenting yourself, I submit my first bunch of comments now... :)

falbrechtskirchinger commented on doc/mkdocs/docs/css/custom.css:
```diff
@@ -0,0 +1,4 @@
+/* disable ligatures in code and preformatted blocks */
+code, pre {
+    font-variant-ligatures: none;
+}
```
The chosen mkdocs font supports ligatures and messes with `<=>`, etc.

nlohmann replied:
You mean like

<img width="426" alt="image" src="https://user-images.githubusercontent.com/159488/164782229-1de6910b-7ac9-4ce0-8633-0830a1e9bd81.png">

?

falbrechtskirchinger replied:
Exactly.

falbrechtskirchinger commented on doc/mkdocs/mkdocs.yml:
```diff
@@ -31,6 +31,7 @@ theme:
 nav:
   - Home:
     - index.md
+    - "API Documentation": api/basic_json/index.md
```
Remove.

falbrechtskirchinger commented on doc/mkdocs/mkdocs.yml:
```diff
@@ -150,6 +151,7 @@ nav:
         - 'operator value_t': api/basic_json/operator_value_t.md
         - 'operator[]': api/basic_json/operator[].md
         - 'operator=': api/basic_json/operator=.md
+        - 'operator<=>': api/basic_json/operator_spaceship.md
```
The order of these operators is highly unintuitive. Should we order them strictly lexicographically, logically or a (better) mix of both? For example, `operator<<` is interspersed with `operator<`, etc.

nlohmann replied:
You're right. I guess the best order could be the one on https://json.nlohmann.me/api/basic_json/

falbrechtskirchinger replied:
I'm trying a new order based on that.

falbrechtskirchinger commented on include/nlohmann/detail/iterators/iteration_proxy.hpp:
```diff
@@ -122,25 +149,34 @@ template<typename IteratorType> class iteration_proxy
 
     /// return iterator begin (needed for range-based for)
-    iteration_proxy_value<IteratorType> begin() noexcept
+    iteration_proxy_value<IteratorType> begin() const noexcept
```
This is something David Stone mentioned and I wanted to read his comment again, to see if I missed anything.

falbrechtskirchinger commented on include/nlohmann/detail/iterators/iteration_proxy.hpp:
```diff
@@ -122,25 +149,34 @@ template<typename IteratorType> class iteration_proxy
   private:
     /// the container to iterate
-    typename IteratorType::reference container;
+    typename IteratorType::pointer container;
```
Double check if changing this to a pointer has any implications elsewhere in the code. Specifically, I remember an assert that may have a different meaning now. (Also set to `nullptr`.)

falbrechtskirchinger commented on include/nlohmann/detail/iterators/iteration_proxy.hpp:
```diff
@@ -187,3 +223,8 @@ class tuple_element<N, ::nlohmann::detail::iteration_proxy_value<IteratorType >>
+#if JSON_HAS_RANGES
+    template <typename IteratorType>
+    inline constexpr bool ::std::ranges::enable_borrowed_range<::nlohmann::detail::iteration_proxy<IteratorType>> = true;
+#endif
```
I assume modern MSVC can handle `<::nlohmann` now?

nlohmann replied:
I have no idea... MSVC is (to me) always good for a surprise...

falbrechtskirchinger replied:
I think this was related to digraphs/trigraphs which have been removed in C++17? So it shouldn't be a concern in a C++20 code path. But as you say, it is MSVC we're talking about ... ;-)

falbrechtskirchinger commented on include/nlohmann/detail/macro_scope.hpp:
```diff
@@ -98,14 +98,39 @@
+        #if __has_include(<compare>)
+            #include <compare>
+            #if defined(__cpp_lib_three_way_comparison) && __cpp_lib_three_way_comparison >= 201907L
+                #define JSON_HAS_THREE_WAY_COMPARISON 1
```
`#if __has_include(<compare>)` needs to be on its own. If merged with the line above, MSVC's preprocessor isn't happy about the angle brackets.

falbrechtskirchinger commented on include/nlohmann/detail/macro_unscope.hpp:
```diff
@@ -25,6 +25,7 @@
     #undef JSON_HAS_FILESYSTEM
     #undef JSON_HAS_EXPERIMENTAL_FILESYSTEM
     #undef JSON_HAS_THREE_WAY_COMPARISON
+    #undef JSON_HAS_RANGES
```
`JSON_LEGACY_COMPARISON` is missing?

nlohmann replied:
Seems so.

nlohmann commented on include/nlohmann/detail/value_t.hpp:
```diff
@@ -64,7 +69,11 @@ Returns an ordering that is similar to Python:
 */
-inline bool operator<(const value_t lhs, const value_t rhs) noexcept
+#if JSON_HAS_THREE_WAY_COMPARISON
+    inline std::partial_ordering operator<=>(const value_t lhs, const value_t rhs) noexcept // *NOPAD*
```
I understand. I guess it's time to move on to Clang-Format - but then I need the active PRs to be merged before to avoid bigger problems.

falbrechtskirchinger commented on test/src/unit-class_parser.cpp:
```diff
@@ -33,6 +33,7 @@ SOFTWARE.
 #include <nlohmann/json.hpp>
 using nlohmann::json;
 
+#include <iostream>
```
Left over from debugging, I guess? Remove.

falbrechtskirchinger commented:
> Since I keep making comments to things you are then commenting yourself, I submit my first bunch of comments now... :)

Sorry about that. :-)

nlohmann commented on test/src/unit-regression2.cpp:
```diff
@@ -785,6 +791,7 @@ TEST_CASE("regression tests 2")
     }
 
 #if JSON_HAS_FILESYSTEM || JSON_HAS_EXPERIMENTAL_FILESYSTEM
+    // JSON_HAS_CPP_17 (do not remove; see note at top of file)
```
I begin to wonder whether it would make more sense to teach CMake to know that `JSON_HAS_FILESYSTEM` implies `JSON_HAS_CPP_17` rather than to add magic words and justifying remarks.

falbrechtskirchinger replied:
I thought about that as well. Either option seems at least a little bit messy. I'll see what it would take.

nlohmann replied:
At least it would be messy at one place rather than spread around multiple unit tests.

falbrechtskirchinger replied:
Fair point. :-)

falbrechtskirchinger replied:
Added to the list.

nlohmann commented on doc/mkdocs/mkdocs.yml:
```diff
@@ -150,15 +150,16 @@ nav:
+        - 'operator+=': api/basic_json/operator+=.md
         - 'operator==': api/basic_json/operator_eq.md
         - 'operator!=': api/basic_json/operator_ne.md
         - 'operator<': api/basic_json/operator_lt.md
```
I think we should either have

- `<`, `>`, `<=`, `>=` (like cppreference) or
- `<`, `<=`, `>`, `>=` (similar to the order before).

But

- `<`, `<=`, `>=`, `>`

feels odd to me.

falbrechtskirchinger replied:
Hm.
less than, less than or equal, (equal), greater than or equal, greater than
Feels natural to me but I guess this is going to be subjective either way. Might as well pick the cppreference order then.

nlohmann replied:
Yes, pleae.

nlohmann commented on include/nlohmann/detail/iterators/iteration_proxy.hpp:
```diff
@@ -6,6 +6,10 @@
 #include <utility> // move
 
+#if JSON_HAS_RANGES
+    #include <ranges> // enable_borrowed_range
```
I think it would be a good idea to add some examples for using ranges to the API page of `items`. What do you think?

falbrechtskirchinger replied:
There're a few place where examples would be helpful. I was being lazy. ;-)

But ranges are tricky anyway because of the mixed and at times sparse support in the big standard library implementations. The unit test is similarly meager because testing and guarding each algorithm/view is time consuming.
Then there's the incompatibility between the GNU standard library and Clang.

nlohmann replied:
I totally understand! Then let's not overwhelm this PR and have the examples as homework for later...

nlohmann commented on doc/mkdocs/docs/api/basic_json/value_t.md:
```diff
@@ -22,12 +22,43 @@ functions [`is_null`](is_null.md), [`is_object`](is_object.md), [`is_array`](is_
 [`is_discarded`](is_discarded.md), [`is_binary`](is_binary.md), [`is_primitive`](is_primitive.md), and
 [`is_structured`](is_structured.md) rely on it.
 
+## Ordering
```
Can you move this to the Notes section? We have standardized sections throughout the documentation (see check_structure.py).

nlohmann commented on doc/mkdocs/docs/api/basic_json/value_t.md:
```diff
@@ -22,12 +22,43 @@ functions [`is_null`](is_null.md), [`is_object`](is_object.md), [`is_array`](is_
-distinguishes these three types for numbers: [`number_unsigned_t`](number_unsigned_t.md) is used for unsigned integers,
-[`number_integer_t`](number_integer_t.md) is used for signed integers, and [`number_float_t`](number_float_t.md) is used
-for floating-point numbers or to approximate integers which do not fit in the limits of their respective type.
+!!! note
```
Please add a title - like "Number handling".

nlohmann commented on doc/mkdocs/docs/api/basic_json/operator_spaceship.md:
```diff
@@ -0,0 +1,70 @@
+
+## Notes
+
+!!! note
```
Add title "NaN handling".

nlohmann commented on doc/mkdocs/docs/api/basic_json/operator_ne.md:
```diff
@@ -37,6 +53,16 @@ No-throw guarantee: this function never throws exceptions.
 
+## Notes
+
+!!! note
```
Add title "NaN handling".

nlohmann commented on doc/mkdocs/docs/api/basic_json/operator_lt.md:
```diff
@@ -52,6 +55,18 @@ No-throw guarantee: this function never throws exceptions.
 
+## Notes
+
+!!! note
```
Add title "NaN handling".

nlohmann commented on doc/mkdocs/docs/api/basic_json/operator_le.md:
```diff
@@ -38,6 +46,18 @@ No-throw guarantee: this function never throws exceptions.
 
+## Notes
+
+!!! note
```
Add title "NaN handling".

nlohmann commented on doc/mkdocs/docs/api/basic_json/operator_gt.md:
```diff
@@ -37,6 +45,18 @@ No-throw guarantee: this function never throws exceptions.
 
+## Notes
+
+!!! note
```
Add title "NaN handling".

nlohmann commented on doc/mkdocs/docs/api/basic_json/operator_ge.md:
```diff
@@ -38,6 +46,18 @@ No-throw guarantee: this function never throws exceptions.
 
+## Notes
+
+!!! note
```
Add title "NaN handling".

nlohmann commented on doc/mkdocs/docs/api/basic_json/operator_eq.md:
```diff
@@ -46,7 +56,11 @@ Linear.
 
 !!! note
```
Add title "NaN handling".

falbrechtskirchinger replied:
How about "Handling NaN and special values"?

I'm trying to reconcile the explanations about `null` and discarded in the same block. Although, `null` isn't particularly special...
Close enough?

nlohmann replied:
What about "Handling NaN and `null`"? Or do we have special values beyond?

(I think `null` is somewhat special)

falbrechtskirchinger replied:
`NaN`, `null`, and discarded.

"Handling of `NaN`, `null`, and discarded values"?

Also, does "handling" accurately capture what's being talked about in this note?

How about:
(1) "Comparison of `NaN`, `null`, and discarded values" or (2) simply "Comparison of special values".

(In the other block(s) we have `NaN` and an explanation about the rewritten candidate operator, which I've split into a separate note.)

nlohmann replied:
"Comparison of special values" 👍

nlohmann commented on doc/mkdocs/docs/api/macros/json_legacy_comparison.md:
```diff
@@ -0,0 +1,56 @@
+      - The `<=>` operator will always give the correct result (`#!cpp false`).
+      - Overloads for the equality and relational operators emulate the legacy behavior.
+
+    Code outside your control may use either the 3-way comparison or the equality and
```
Maybe a bit nitpicky, but mixing "<=>" (above) and "3-way comparison" (here) may confuse readers new to C++20.

nlohmann commented on doc/mkdocs/docs/api/macros/json_legacy_comparison.md:
```diff
@@ -0,0 +1,56 @@
+    When targeting C++20 or above, enabling the legacy comparison behavior is _strongly_
+    discouraged.
+
+      - The `<=>` operator will always give the correct result (`#!cpp false`).
```
... regardless of the value of `JSON_LEGACY_COMPARISON`.

falbrechtskirchinger commented:
@nlohmann Could you please comment on the two open questions in the to-do list at the bottom of the PR description?

gregmarr commented:
> NaN values always compare false/unordered when compared to themselves, other NaNs, or other numbers.

Is the legacy behavior `JSON_LEGACY_COMPARISON=1` not included for pre-C++20 because it is using the same code as `<=>` with the macros for implementing the operators?

falbrechtskirchinger commented:
> > NaN values always compare false/unordered when compared to themselves, other NaNs, or other numbers.
> 
> Is the legacy behavior `JSON_LEGACY_COMPARISON=1` not included for pre-C++20 because it is using the same code as `<=>` with the macros for implementing the operators?

No, it's because that's the conclusion we reached here: https://github.com/nlohmann/json/issues/3409#issuecomment-1091661041

The NaN behavior is deliberately fixed in all three code paths without an ability to opt-out.

gregmarr commented:
Great, thanks for the explanation.  Agree 100%.

nlohmann commented:
> @nlohmann Could you please comment on the two open questions in the to-do list at the bottom of the PR description?

My take on the questions:

**Should JSON_LEGACY_COMPARISON=1 be added to CI?**

Yes.

**Should JSON_LEGACY_COMPARISON be renamed to JSON_LEGACY_DISCARDED_COMPARISON (or something else entirely) now that it only applies to discarded values?**

Yes - the name `JSON_LEGACY_DISCARDED_COMPARISON` (though quite cumbersome) is still better than `JSON_LEGACY_COMPARISON`.

@gregmarr @falbrechtskirchinger What do you think?

gregmarr commented:
Perhaps `JSON_DISCARDED_VALUE_USE_LEGACY_COMPARISON`.

gregmarr commented:
>Should JSON_LEGACY_COMPARISON=1 be added to CI?

Do you mean that we should have tests both with and without it?  If so, yes.

falbrechtskirchinger commented:
> > @nlohmann Could you please comment on the two open questions in the to-do list at the bottom of the PR description?
> 
> My take on the questions:
> 
> **Should JSON_LEGACY_COMPARISON=1 be added to CI?**
> 
> Yes.

Ok. I don't think I've added a CMake option yet. That would conveniently go along with that.

> 
> **Should JSON_LEGACY_COMPARISON be renamed to JSON_LEGACY_DISCARDED_COMPARISON (or something else entirely) now that it only applies to discarded values?**
> 
> Yes - the name `JSON_LEGACY_DISCARDED_COMPARISON` (though quite cumbersome) is still better than `JSON_LEGACY_COMPARISON`.
> 
> @gregmarr @falbrechtskirchinger What do you think?

I'm for a more descriptive name, yes.

> Perhaps `JSON_DISCARDED_VALUE_USE_LEGACY_COMPARISON`.

If we go that route, it should be more in line with `JSON_USE_IMPLICIT_CONVERSIONS`:

`JSON_USE_LEGACY_DISCARDED_VALUE_COMPARISON`

Of course, that really doesn't fit in the sidebar anymore, does it? Hm.

nlohmann commented:
`JSON_USE_LEGACY_DISCARDED_VALUE_COMPARISON` is fine with me, too. And don't worry about the sidebar...

falbrechtskirchinger commented:
I discovered that the way `JSON_HAS_THREE_WAY_COMPARISON` and `JSON_HAS_RANGES` are detected is still flawed. Need to test and then back port the fix from the `apply` branch/PR.

falbrechtskirchinger commented:
This should wait for #3229. There is a known/suspected issue with the Intel compiler.

nlohmann commented:
> This should wait for #3229. There is a known/suspected issue with the Intel compiler.

Merged now (https://github.com/nlohmann/json/pull/3465).

falbrechtskirchinger commented:
The unit tests do not trigger the behavior I'm observing in my synthetic test case on Compiler Explorer. Without a failure, there's nothing to fix. :man_shrugging:

nlohmann commented:
So: ready for a final review?

falbrechtskirchinger commented:
Yes.

falbrechtskirchinger commented on tests/src/unit-comparison.cpp:
```diff
@@ -164,115 +391,199 @@ TEST_CASE("lexicographical comparison operators")
+                    }
+                    else
+                    {
+                        // check definition
```
Shift around and/or clarify similar comments.

falbrechtskirchinger reviewed (commented):
Focused on code over documentation in this round and found a few minor things.

falbrechtskirchinger commented:
I'm inclined to fix #3472 as well. It's just a bunch of `inline`s for the `to_/from_json` functions. But I'd like to wait a little longer to see if there're any other problems given that GCC segfaults with those changes at a later stage.

falbrechtskirchinger commented:
Based on further testing and the GCC bug report, I do not believe that the modules fix should cause any problems. I've included it with the last update.

falbrechtskirchinger commented on tests/src/unit-comparison.cpp:
```diff
@@ -158,121 +388,225 @@ TEST_CASE("lexicographical comparison operators")
+                        // as the inverse of less than or equal which is defined as
+                        // the inverse of less than with the operand order reversed
+                        CHECK((j_values[i] > j_values[j]) == !(j_values[i] <= j_values[j]));
+                        CHECK((j_values[i] > j_values[j]) == !!(j_values[j] < j_values[i]));
```
@nlohmann This needs to be ignored in Codacy.

nlohmann replied:
The issue should be ignored in the next run.

nlohmann commented:
@falbrechtskirchinger Is there anything I can do to help here?

falbrechtskirchinger commented:
> @falbrechtskirchinger Is there anything I can do to help here?

Yes. Please review and merge. :-)

I see a minor merge conflict in the documentation that I can resolve. Otherwise, this PR is done.

gregmarr commented:
There's also a Codacy issue that needs to be ignored in the comparison tests.

nlohmann commented:
> There's also a Codacy issue that needs to be ignored in the comparison tests.

I marked it as ignored.

nlohmann reviewed (approved):
Looks good to me.

nlohmann commented:
I would feel more comfortable if @gregmarr could have a second look at this before merging.

falbrechtskirchinger commented:
I'd also like to have one last look at the tables in the unit test and spot-check them.

gregmarr commented on docs/mkdocs/docs/api/basic_json/operator_eq.md:
```diff
@@ -46,7 +56,11 @@ Linear.
 
-    - NaN values never compare equal to themselves or to other NaN values.
+    - `NaN` values are unordered within the domain of numbers.
+      The following comparisons all yield `#!cpp false`:
```
Is it not true that all equality comparisons with a `NaN` value return `false` now?  If that's true, then is there a need to call out specific things which return `false`?

falbrechtskirchinger replied:
No. See https://github.com/nlohmann/json/issues/3409#issuecomment-1100829527 where I asked about this specifically. What you state was initially true, but I realized, that that's nonsensical.

gregmarr replied:
Makes sense, thanks.

gregmarr commented on docs/mkdocs/docs/api/basic_json/operator_ge.md:
```diff
@@ -1,17 +1,25 @@
+1. Compares whether one JSON value `lhs` is greater than or equal to another JSON value `rhs`
+   according to the following rules:
+    - The comparison always yields `#!cpp false` if (1) either operand is discarded, or (2) either
+      operand is `NaN` and the other operand is either `NaN` or any other number.
```
What do we get for `NaN >= "string"`?  I assume that this would be a "not the same type" comparison, unlike `discarded >= "string"` where it's always false.

falbrechtskirchinger replied:
Is that not adequately explained by the next line?

falbrechtskirchinger replied:
Looking at it again, there's room for improvement. Added to my to-do list for later.

gregmarr commented on docs/mkdocs/docs/api/basic_json/operator_spaceship.md:
```diff
@@ -0,0 +1,70 @@
+    - Two JSON values compare with a result of `std::partial_ordering::unordered` if either value is discarded.
+    - If both JSON values are of the same type, the result is produced by 3-way comparing their stored values using their
+      respective `operator<=>`.
+    - Integer and floating-point numbers are converted to their common type and then 3-way compared using their respective
```
Would this (and other related docs) be better as `integer numbers are converted to the floating point number type when compared with a floating point number` or is this also covering `signed integer` to `unsigned integer` comparisons?

falbrechtskirchinger replied:
It's covering signed and unsigned integers, yes.

gregmarr replied:
Thanks.

gregmarr commented on include/nlohmann/detail/conversions/from_json.hpp:
```diff
@@ -482,11 +482,18 @@ struct from_json_fn
 namespace // NOLINT(cert-dcl59-cpp,fuchsia-header-anon-namespaces,google-build-namespaces)
 {
+#endif
+#ifdef JSON_HAS_CPP_17
```
Should this `#endif/#ifdef` be an `#else`?

falbrechtskirchinger replied:
No. As intended.

falbrechtskirchinger replied:
I have to push a small cosmetic fix anyway, so I'm making this change as well.

gregmarr commented on include/nlohmann/detail/conversions/to_json.hpp:
```diff
@@ -420,11 +420,18 @@ struct to_json_fn
 namespace // NOLINT(cert-dcl59-cpp,fuchsia-header-anon-namespaces,google-build-namespaces)
 {
+#endif
+#ifdef JSON_HAS_CPP_17
```
#else as above

falbrechtskirchinger replied:
Same.

gregmarr reviewed (commented):
A couple minor questions/observations, otherwise LGTM.

falbrechtskirchinger commented:
I've fixed a misaligned column numbers comment in the unit test, and added `JSON_INLINE_VARIABLE`.

I've also completed my spot-check of the unit test tables and added a `CHECK()` for `std::is_gt()` to the cross-check loops. I'm as certain that these tables are correct as I'm ever going to be.

@nlohmann Ready for merge pending all-green CI.
(There's an opportunity to clarify how different types compare on the individual operator pages. I spotted at least two very minor issues in the BJData documentation but missed my window to leave a comment. I'll submit a separate PR with various small doc fixes later.)

nlohmann commented:
Thanks a lot!!!

gregmarr commented:
@falbrechtskirchinger I wonder if this is the reason for the issues we had with differing compiler behaviors: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2468r2.html

falbrechtskirchinger commented:
> @falbrechtskirchinger I wonder if this is the reason for the issues we had with differing compiler behaviors: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2468r2.html

Interesting find. I just briefly scanned through it. Based on that, my initial thought is no. Our issue is specific to enums and how overload resolution is specified for them. I wish I knew what the proper process is to bring (perceived) language defects to the attention of the standard committee. Maybe I should just email the authors of that paper.
